File: crates/cnab-core/src/condicao.rs
//! Avaliação dos arquétipos da DSL sobre o arquivo.
//!
//! `None` significa **não avaliável** — é o que acontece com `custom` e com o que
//! depende de algo que o spec não carrega. Nunca devolver `Some(false)` nesses
//! casos é deliberado: regra silenciosamente aprovada cria falsa confiança.

use cnab_specs::{Condicao, ModoComparacao, Parcela, Posicao, SentidoDominio, VariavelDaGuarda};

use crate::expressao::{Contexto, aplicar_replace, substring};
use crate::valor::{Valor, comparar};

pub fn avaliar_condicao(condicao: &Condicao, ctx: &Contexto) -> Option<bool> {
    match condicao {
        Condicao::LiteralFixo {
            alvo,
            posicao,
            operador,
            valor,
            comparacao,
        } => {
            let campo = ler_faixa(alvo, posicao, ctx)?;
            comparar_literal(operador, &campo, valor, *comparacao)
        }

        Condicao::Dominio {
            alvo,
            posicao,
            valores,
            sentido,
            comparacao,
        } => {
            let campo = ler_faixa(alvo, posicao, ctx)?;
            let mut casa_algum = false;
            for valor in valores {
                casa_algum |= comparar_literal("==", &campo, valor, *comparacao)?;
            }
            // `permitidos` vem de conjunção de desigualdades: o erro é não casar
            // nenhum. `proibidos` vem de disjunção de igualdades: o erro é casar.
            Some(match sentido {
                SentidoDominio::Permitidos => !casa_algum,
                SentidoDominio::Proibidos => casa_algum,
            })
        }

        Condicao::Intervalo {
            alvo,
            posicao,
            limites,
            comparacao,
        } => {
            let campo = ler_faixa(alvo, posicao, ctx)?;
            let mut todos = true;
            for limite in limites {
                todos &= comparar_literal(&limite.operador, &campo, &limite.valor, *comparacao)?;
            }
            Some(todos)
        }

        Condicao::NumericoBranco {
            alvo,
            posicao,
            exige: _,
            residuo,
        } => {
            let campo = ler_faixa(alvo, posicao, ctx)?;
            let resto = aplicar_replace(&campo, &residuo.padrao, "").ok()?;
            let residual = Valor::Numero(resto.chars().count() as f64);
            let numerico_invalido = Valor::texto(&campo).como_numero().is_nan();
            let residual_falha = comparar(
                &residuo.operador,
                &residual,
                &Valor::Numero(residuo.valor as f64),
            )?;
            Some(numerico_invalido || residual_falha)
        }

        Condicao::CoerenciaRegistro {
            alvo,
            posicao,
            operador,
            outro,
            posicao_outro,
        } => {
            let campo = ler_faixa(alvo, posicao, ctx)?;
            let outro = ler_faixa(outro, posicao_outro, ctx)?;
            comparar(operador, &Valor::Texto(campo), &Valor::Texto(outro))
        }

        Condicao::TamanhoLinha {
            alvo,
            operador,
            tamanho,
        } => {
            let indice = resolver_indice(alvo, ctx)?;
            let comprimento = ctx.linha(indice).chars().count();
            comparar(
                operador,
                &Valor::Numero(comprimento as f64),
                &Valor::Numero(*tamanho as f64),
            )
        }

        Condicao::Modulo11 {
            alvo,
            posicao,
            base,
            modulo,
            resultado,
            transformacao,
            ..
        } => {
            // A função que o fonte aplica à faixa antes de comparar não está no
            // spec — sem ela, calcular seria inventar um resultado.
            if transformacao.is_some() {
                return None;
            }
            let esperado = calcular_digito(base, *modulo, resultado, ctx)?;
            let digito = ler_faixa(alvo, posicao, ctx)?;
            // O fonte compara sem aspas contra número e com aspas contra letra;
            // `==` cobre os dois com a mesma coerção que ele usa.
            Some(!crate::valor::igual_js(&Valor::Texto(digito), &esperado))
        }

        Condicao::Disjuncao { partes, .. } => {
            let mut algum = false;
            for parte in partes {
                algum |= avaliar_condicao(parte, ctx)?;
            }
            Some(algum)
        }

        Condicao::Conjuncao { partes, .. } => {
            let mut todos = true;
            for parte in partes {
                todos &= avaliar_condicao(parte, ctx)?;
            }
            Some(todos)
        }

        Condicao::Custom { .. } => None,
    }
}

fn comparar_literal(
    operador: &str,
    campo: &str,
    valor: &str,
    comparacao: ModoComparacao,
) -> Option<bool> {
    // Comparação frouxa é a do fonte contra literal sem aspas: o JavaScript
    // converte a faixa lida para número antes de comparar, e é isso que faz
    // `" 1"` passar como `01` e o campo em branco valer zero.
    let direita = match comparacao {
        ModoComparacao::Frouxa => Valor::Numero(crate::valor::numero_js(valor)),
        ModoComparacao::Estrita => Valor::texto(valor),
    };
    comparar(operador, &Valor::texto(campo), &direita)
}

/// Resto da soma ponderada, primeira metade do cálculo do dígito. O fonte também
/// o compara direto, sem virar dígito, para escolher qual regra aplicar.
pub fn calcular_resto(base: &[Parcela], modulo: i64, ctx: &Contexto) -> Option<f64> {
    if modulo == 0 {
        return None;
    }
    // A função que o fonte aplica a cada parcela do CNPJ alfanumérico não está no
    // spec — sem ela, calcular seria inventar um resultado.
    if base.iter().any(|p| p.transformacao.is_some()) {
        return None;
    }

    let mut soma = 0.0;
    for parcela in base {
        let campo = ler_faixa(
            &parcela.alvo,
            &Posicao {
                inicio0: parcela.inicio0,
                fim0: parcela.fim0,
            },
            ctx,
        )?;
        soma += Valor::texto(campo).como_numero() * parcela.peso as f64;
    }
    if soma.is_nan() {
        return None;
    }

    Some(soma % modulo as f64)
}

pub fn calcular_digito(
    base: &[Parcela],
    modulo: i64,
    resultado: &[cnab_specs::FaixaDeResto],
    ctx: &Contexto,
) -> Option<Valor> {
    let resto = calcular_resto(base, modulo, ctx)?;

    // A ordem do fonte é a ordem de avaliação: a última atribuição que casa vence.
    let mut esperado: Option<Valor> = None;
    for faixa in resultado {
        let casa = match (&faixa.operador, faixa.resto) {
            (Some(operador), Some(valor)) => comparar(
                operador,
                &Valor::Numero(resto),
                &Valor::Numero(valor as f64),
            )?,
            _ => true,
        };
        if !casa {
            continue;
        }

        if let Some(valor) = &faixa.valor {
            esperado = Some(Valor::texto(valor));
            continue;
        }
        esperado = Some(avaliar_valor(&faixa.expressao, resto)?);
    }

    esperado
}

/// Avalia uma expressão que devolve valor (não booleano), como `11 - resto`.
fn avaliar_valor(fonte: &str, resto: f64) -> Option<Valor> {
    let limpo = fonte.trim();

    if let Some(interno) = literal_entre_aspas(limpo) {
        return Some(Valor::texto(interno));
    }
    if let Ok(inteiro) = limpo.parse::<i64>() {
        return Some(Valor::Numero(inteiro as f64));
    }

    // `11 - resto`: a única forma de expressão que o fonte usa no dígito.
    let (esquerda, direita) = limpo.split_once('-')?;
    let base: f64 = esquerda.trim().parse().ok()?;
    if direita.trim() != "resto" {
        return None;
    }
    Some(Valor::Numero(base - resto))
}

fn literal_entre_aspas(s: &str) -> Option<&str> {
    let bytes = s.as_bytes();
    if bytes.len() >= 2
        && (bytes[0] == b'"' || bytes[0] == b'\'')
        && bytes[bytes.len() - 1] == bytes[0]
        && !s[1..s.len() - 1].contains(bytes[0] as char)
    {
        return Some(&s[1..s.len() - 1]);
    }
    None
}

/// Resolve as variáveis que a guarda referencia. Variável que não se calcula fica
/// de fora, e a guarda que a cita passa a ser recusada por identificador
/// desconhecido — nunca aprovada com um valor inventado.
pub fn resolver_variaveis(variaveis: &[VariavelDaGuarda], ctx: &Contexto) -> Vec<(String, Valor)> {
    let mut resolvidas = Vec::new();
    for variavel in variaveis {
        let valor = match variavel {
            VariavelDaGuarda::Modulo11 {
                base,
                modulo,
                resultado,
                ..
            } => calcular_digito(base, *modulo, resultado, ctx),
            VariavelDaGuarda::Resto { base, modulo, .. } => {
                calcular_resto(base, *modulo, ctx).map(Valor::Numero)
            }
        };
        if let Some(valor) = valor {
            resolvidas.push((variavel.nome().to_string(), valor));
        }
    }
    resolvidas
}

fn ler_faixa(alvo: &str, posicao: &Posicao, ctx: &Contexto) -> Option<String> {
    let indice = resolver_indice(alvo, ctx)?;
    Some(substring(ctx.linha(indice), posicao.inicio0, posicao.fim0))
}

/// `res[i]`, `res[0]`, `res[i + 2]`, `res[j]` — o índice pode ser aritmético.
pub fn resolver_indice(alvo: &str, ctx: &Contexto) -> Option<i64> {
    let dentro = alvo.strip_prefix("res[")?.strip_suffix(']')?;
    indice_de(dentro.trim(), ctx)
}

fn indice_de(expressao: &str, ctx: &Contexto) -> Option<i64> {
    if let Ok(numero) = expressao.parse::<i64>() {
        return Some(numero);
    }

    if let Some((nome, resto)) = expressao.split_once(['+', '-']) {
        let operador = expressao.as_bytes()[nome.len()];
        let base = variavel_de_indice(nome.trim(), ctx)?;
        let passo: i64 = resto.trim().parse().ok()?;
        return Some(if operador == b'+' {
            base + passo
        } else {
            base - passo
        });
    }

    variavel_de_indice(expressao, ctx)
}

fn variavel_de_indice(nome: &str, ctx: &Contexto) -> Option<i64> {
    match nome {
        "i" => Some(ctx.i as i64),
        // `j = i + 1` no fonte.
        "j" => Some(ctx.i as i64 + 1),
        outro => match ctx.variaveis.get(outro) {
            Some(Valor::Numero(n)) => Some(*n as i64),
            _ => None,
        },
    }
}

File: tools/spec-extractor/src/runner/condicao.ts
import type { DslCondition } from "../rule-mapper.js";
import {
  aplicarComparacao,
  avaliarExpressao,
  ExpressaoNaoSuportada,
  linhaDe,
  type ContextoArquivo,
  type Valor,
} from "./expressao.js";

/**
 * Avalia um arquétipo da DSL sobre o arquivo. `null` significa **não avaliável** —
 * é o que acontece com `custom` e com o que depende de coisa que o spec não
 * carrega. Nunca devolver `false` nesses casos é deliberado: regra silenciosamente
 * aprovada cria falsa confiança.
 */
export function avaliarCondicao(
  condicao: DslCondition,
  ctx: ContextoArquivo
): boolean | null {
  switch (condicao.tipo) {
    case "literal_fixo": {
      const campo = lerFaixa(condicao.alvo, condicao.posicao, ctx);
      if (campo === null) return null;
      return comparar(condicao.operador, campo, condicao.valor, condicao.comparacao);
    }

    case "dominio": {
      const campo = lerFaixa(condicao.alvo, condicao.posicao, ctx);
      if (campo === null) return null;
      const casaAlgum = condicao.valores.some((valor) =>
        comparar("==", campo, valor, condicao.comparacao)
      );
      // `permitidos` vem de uma conjunção de desigualdades: o erro é não casar
      // nenhum. `proibidos` vem de uma disjunção de igualdades: o erro é casar.
      return condicao.sentido === "permitidos" ? !casaAlgum : casaAlgum;
    }

    case "intervalo": {
      const campo = lerFaixa(condicao.alvo, condicao.posicao, ctx);
      if (campo === null) return null;
      return condicao.limites.every((limite) =>
        comparar(limite.operador, campo, limite.valor, condicao.comparacao)
      );
    }

    case "numerico_branco": {
      const campo = lerFaixa(condicao.alvo, condicao.posicao, ctx);
      if (campo === null) return null;
      const residuo = campo.replace(new RegExp(condicao.residuo.padrao, "g"), "").length;
      return (
        isNaN(Number(campo)) ||
        aplicarComparacao(condicao.residuo.operador, residuo, condicao.residuo.valor)
      );
    }

    case "coerencia_registro": {
      const campo = lerFaixa(condicao.alvo, condicao.posicao, ctx);
      const outro = lerFaixa(condicao.outro, condicao.posicao_outro, ctx);
      if (campo === null || outro === null) return null;
      return aplicarComparacao(
        condicao.operador,
        deslocar(campo, condicao.ajuste),
        deslocar(outro, condicao.ajuste_outro)
      );
    }

    case "numero_da_linha": {
      const campo = lerFaixa(condicao.alvo, condicao.posicao, ctx);
      if (campo === null) return null;
      let numero: number;
      try {
        numero = indiceDe(condicao.fluxo, ctx);
      } catch (erro) {
        if (erro instanceof ExpressaoNaoSuportada) return null;
        throw erro;
      }
      // O fonte compara texto com número, e o `==` do JavaScript coage a faixa:
      // `"000006"` passa como 6.
      return aplicarComparacao(condicao.operador, campo, numero);
    }

    case "tamanho_linha": {
      const indice = resolverIndice(condicao.alvo, ctx);
      if (indice === null) return null;
      return aplicarComparacao(
        condicao.operador,
        linhaDe(ctx, indice).length,
        condicao.tamanho
      );
    }

    case "modulo_11":
      return avaliarModulo11(condicao, ctx);

    case "disjuncao": {
      const partes = condicao.partes.map((parte) => avaliarCondicao(parte, ctx));
      if (partes.some((p) => p === null)) return null;
      return partes.some(Boolean);
    }

    case "conjuncao": {
      const partes = condicao.partes.map((parte) => avaliarCondicao(parte, ctx));
      if (partes.some((p) => p === null)) return null;
      return partes.every(Boolean);
    }

    case "custom":
      return null;
  }
}

/**
 * Resto da soma ponderada, que é a primeira metade do cálculo do dígito. O fonte
 * também o compara direto, sem virar dígito, para escolher qual regra aplicar.
 */
export function calcularResto(
  calculo: {
    base: Extract<DslCondition, { tipo: "modulo_11" }>["base"];
    modulo: number;
    /** Redução por excesso do módulo 10; ausente é a soma ponderada direta. */
    dobra?: { limite: number; subtrai: number } | null;
  },
  ctx: ContextoArquivo
): number | null {
  // A função que o fonte aplica a cada parcela do CNPJ alfanumérico não está no
  // spec — sem ela, calcular seria inventar um resultado.
  if (calculo.base.some((parcela) => parcela.transformacao !== null)) return null;

  const dobra = calculo.dobra ?? null;
  let soma = 0;
  for (const parcela of calculo.base) {
    const campo = lerFaixa(parcela.alvo, parcela, ctx);
    if (campo === null) return null;
    const produto = Number(campo) * parcela.peso;
    // O fonte reduz parcela a parcela, antes de somar: reduzir o total daria
    // outro número. É o `if (faixa * 2 > 9) soma = (faixa * 2) - 9` do módulo 10.
    soma += dobra && produto > dobra.limite ? produto - dobra.subtrai : produto;
  }
  if (isNaN(soma)) return null;

  return soma % calculo.modulo;
}

/**
 * O dígito que o fonte calcula: soma ponderada, resto da divisão, e o valor
 * esperado por faixa de resto. Vale tanto para a condição da regra quanto para a
 * variável que a guarda referencia — é o mesmo cálculo, escrito duas vezes no
 * fonte porque ele repete o bloco por dígito informado.
 */
export function calcularDigito(
  calculo: {
    base: Extract<DslCondition, { tipo: "modulo_11" }>["base"];
    modulo: number;
    dobra?: { limite: number; subtrai: number } | null;
    resultado: Extract<DslCondition, { tipo: "modulo_11" }>["resultado"];
  },
  ctx: ContextoArquivo
): Valor | null {
  const resto = calcularResto(calculo, ctx);
  if (resto === null) return null;

  // A ordem do fonte é a ordem de avaliação: a última atribuição que casa vence.
  let esperado: Valor | null = null;
  for (const faixa of calculo.resultado) {
    const casa =
      faixa.operador === null || faixa.resto === null
        ? true
        : aplicarComparacao(faixa.operador, resto, faixa.resto);
    if (!casa) continue;

    if (faixa.valor !== null) {
      esperado = faixa.valor;
      continue;
    }
    try {
      esperado = avaliarValor(faixa.expressao, { ...ctx, variaveis: { resto } });
    } catch (erro) {
      if (erro instanceof ExpressaoNaoSuportada) return null;
      throw erro;
    }
  }
  return esperado;
}

function avaliarModulo11(
  condicao: Extract<DslCondition, { tipo: "modulo_11" }>,
  ctx: ContextoArquivo
): boolean | null {
  if (condicao.transformacao !== null) return null;

  const esperado = calcularDigito(condicao, ctx);
  if (esperado === null) return null;

  const digito = lerFaixa(condicao.alvo, condicao.posicao, ctx);
  if (digito === null) return null;
  // O fonte compara sem aspas contra número e com aspas contra letra; `==` cobre
  // os dois casos com a mesma coerção que ele usa.
  return !aplicarComparacao("==", digito, esperado);
}

/** Avalia uma expressão que devolve valor (não booleano), como `11 - resto`. */
function avaliarValor(fonte: string, ctx: ContextoArquivo): Valor {
  const limpo = fonte.trim();
  const literal = limpo.match(/^(?:"([^"]*)"|'([^']*)'|(-?\d+))$/);
  if (literal) return literal[1] ?? literal[2] ?? Number(literal[3]);

  const subtracao = limpo.match(/^(\d+)\s*-\s*([A-Za-z_$][\w$]*)$/);
  if (subtracao) {
    const variavel = ctx.variaveis?.[subtracao[2]];
    if (typeof variavel !== "number") {
      throw new ExpressaoNaoSuportada(`variável não resolvida: ${subtracao[2]}`);
    }
    return Number(subtracao[1]) - variavel;
  }

  // `avaliarExpressao` devolve booleano; aqui o que se quer é o valor.
  throw new ExpressaoNaoSuportada(`expressão de valor não suportada: ${limpo}`);
}

/**
 * Aplica o deslocamento do fonte a uma faixa. Sem deslocamento a faixa segue
 * string, e a comparação é textual; com deslocamento o `-` do JavaScript já
 * converteu para número antes do operador ver os dois lados — inclusive quando a
 * conversão dá `NaN`, que é como o fonte reprova faixa não numérica.
 */
function deslocar(campo: string, ajuste: number | null): Valor {
  return ajuste === null ? campo : Number(campo) + ajuste;
}

function comparar(
  operador: string,
  campo: string,
  valor: string,
  comparacao: "estrita" | "frouxa"
): boolean {
  // Comparação frouxa é a do fonte contra literal sem aspas: o JavaScript converte
  // a string lida para número antes de comparar, e é isso que faz `" 1"` passar
  // como `01` e o campo em branco valer zero.
  const direita: Valor = comparacao === "frouxa" ? Number(valor) : valor;
  return aplicarComparacao(operador, campo, direita);
}

function lerFaixa(
  alvo: string,
  posicao: { inicio0: number; fim0: number },
  ctx: ContextoArquivo
): string | null {
  const indice = resolverIndice(alvo, ctx);
  if (indice === null) return null;
  return linhaDe(ctx, indice).substring(posicao.inicio0, posicao.fim0);
}

/** `res[i]`, `res[0]`, `res[i + 2]`, `res[j]` — o índice pode ser aritmético. */
export function resolverIndice(alvo: string, ctx: ContextoArquivo): number | null {
  const dentro = alvo.match(/^res\[(.+)\]$/);
  if (!dentro) return null;
  try {
    return indiceDe(dentro[1], ctx);
  } catch (erro) {
    if (erro instanceof ExpressaoNaoSuportada) return null;
    throw erro;
  }
}

function indiceDe(expressao: string, ctx: ContextoArquivo): number {
  const limpo = expressao.trim();
  const numero = limpo.match(/^\d+$/);
  if (numero) return Number(limpo);

  const aritmetica = limpo.match(/^([A-Za-z_$][\w$]*)\s*([+-])\s*(\d+)$/);
  if (aritmetica) {
    const base = variavelDeIndice(aritmetica[1], ctx);
    const passo = Number(aritmetica[3]);
    return aritmetica[2] === "+" ? base + passo : base - passo;
  }

  const nome = limpo.match(/^[A-Za-z_$][\w$]*$/);
  if (nome) return variavelDeIndice(limpo, ctx);

  throw new ExpressaoNaoSuportada(`índice não suportado: ${limpo}`);
}

function variavelDeIndice(nome: string, ctx: ContextoArquivo): number {
  if (nome === "i") return ctx.i;
  // `j = i + 1` no fonte.
  if (nome === "j") return ctx.i + 1;
  const variavel = ctx.variaveis?.[nome];
  if (typeof variavel === "number") return variavel;
  throw new ExpressaoNaoSuportada(`índice desconhecido: ${nome}`);
}

export { avaliarExpressao };
